// budoux/dist/index.js
"use strict";
/**
 * @license
 * Copyright 2021 Google LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadDefaultParsers = exports.loadDefaultThaiParser = exports.loadDefaultTraditionalChineseParser = exports.loadDefaultSimplifiedChineseParser = exports.loadDefaultJapaneseParser = exports.zhHantModel = exports.zhHansModel = exports.jaModel = exports.HTMLProcessingParser = exports.HTMLProcessor = exports.Parser = void 0;
const ja_js_1 = require("./data/models/ja.js");
Object.defineProperty(exports, "jaModel", { enumerable: true, get: function () { return ja_js_1.model; } });
const zh_hans_js_1 = require("./data/models/zh-hans.js");
Object.defineProperty(exports, "zhHansModel", { enumerable: true, get: function () { return zh_hans_js_1.model; } });
const zh_hant_js_1 = require("./data/models/zh-hant.js");
Object.defineProperty(exports, "zhHantModel", { enumerable: true, get: function () { return zh_hant_js_1.model; } });
const th_js_1 = require("./data/models/th.js");
const html_processor_js_1 = require("./html_processor.js");
var parser_js_1 = require("./parser.js");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return parser_js_1.Parser; } });
var html_processor_js_2 = require("./html_processor.js");
Object.defineProperty(exports, "HTMLProcessor", { enumerable: true, get: function () { return html_processor_js_2.HTMLProcessor; } });
Object.defineProperty(exports, "HTMLProcessingParser", { enumerable: true, get: function () { return html_processor_js_2.HTMLProcessingParser; } });
/**
 * Loads a parser equipped with the default Japanese model.
 * @return A parser with the default Japanese model.
 */
const loadDefaultJapaneseParser = () => {
    return new html_processor_js_1.HTMLProcessingParser(ja_js_1.model);
};
exports.loadDefaultJapaneseParser = loadDefaultJapaneseParser;
/**
 * Loads a parser equipped with the default Simplified Chinese model.
 * @return A parser with the default Simplified Chinese model.
 */
const loadDefaultSimplifiedChineseParser = () => {
    return new html_processor_js_1.HTMLProcessingParser(zh_hans_js_1.model);
};
exports.loadDefaultSimplifiedChineseParser = loadDefaultSimplifiedChineseParser;
/**
 * Loads a parser equipped with the default Traditional Chinese model.
 * @return A parser with the default Traditional Chinese model.
 */
const loadDefaultTraditionalChineseParser = () => {
    return new html_processor_js_1.HTMLProcessingParser(zh_hant_js_1.model);
};
exports.loadDefaultTraditionalChineseParser = loadDefaultTraditionalChineseParser;
/**
 * Loads a parser equipped with the default Thai model.
 * @returns A parser with the default Thai model.
 */
const loadDefaultThaiParser = () => {
    return new html_processor_js_1.HTMLProcessingParser(th_js_1.model);
};
exports.loadDefaultThaiParser = loadDefaultThaiParser;
/**
 * Loads available default parsers.
 * @return A map between available lang codes and their default parsers.
 */
const loadDefaultParsers = () => {
    return new Map([
        ['ja', (0, exports.loadDefaultJapaneseParser)()],
        ['zh-hans', (0, exports.loadDefaultSimplifiedChineseParser)()],
        ['zh-hant', (0, exports.loadDefaultTraditionalChineseParser)()],
        ['th', (0, exports.loadDefaultThaiParser)()],
    ]);
};
exports.loadDefaultParsers = loadDefaultParsers;
